In [2]:
import pandas as pd
import seaborn as sns
import statsmodels.formula.api as smf
import statsmodels.api as sm2
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
from sklearn import metrics 
from sklearn.model_selection  import train_test_split
import numpy as np

%matplotlib inline

In [3]:
df= pd.read_csv('Macro_Economic.csv', header=0)
df.head()

year  quarter   realgdp  realcons  realinv  realgovt  realdpi    cpi  \
0  1959        1  2710.349    1707.4  286.898   470.045   1886.9  28.98   
1  1959        2  2778.801    1733.7  310.859   481.301   1919.7  29.15   
2  1959        3  2775.488    1751.8  289.226   491.260   1916.4  29.35   
3  1959        4  2785.204    1753.7  299.356   484.052   1931.3  29.37   
4  1960        1  2847.699    1770.5  331.722   462.199   1955.5  29.54   

      m1  tbilrate  unemp      pop  infl  realint  
0  139.7      2.82    5.8  177.146  0.00     0.00  
1  141.7      3.08    5.1  177.830  2.34     0.74  
2  140.5      3.82    5.3  178.657  2.74     1.09  
3  140.0      4.33    5.6  179.386  0.27     4.06  
4  139.6      3.50    5.2  180.007  2.31     1.19

In [1]:
s[:5]


NameError: name 's' is not defined

In [2]:
lm= smf.ols(formula='infl ~ year + quarter + realgdp + realcons + realinv + realgovt + realdpi + cpi +  m1 + tbilrate + unemp + pop', data=f).fit()
lm.pvalues

NameError: name 'smf' is not defined

In [29]:
lm1 = smf.ols(formula='infl ~ realgdp + realinv + realgovt + cpi + m1 + tbilrate + pop', data=f).fit()
lm1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   infl   R-squared:                       0.672
Model:                            OLS   Adj. R-squared:                  0.659
Method:                 Least Squares   F-statistic:                     52.56
Date:                Sun, 28 Jun 2020   Prob (F-statistic):           2.79e-40
Time:                        16:19:36   Log-Likelihood:                -376.94
No. Observations:                 188   AIC:                             769.9
Df Residuals:                     180   BIC:                             795.8
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     31.3708     10.172      3.084      0.002      11.299      51.443
realgdp        0.0094      0.001      7.244      0.000       0.007       0.012
realinv       -0.0171      0.003     -6.408      0.000      -0.022      -0.012
realgovt      -0.0132      0.003     -5.019      0.000      -0.018      -0.008
cpi           -0.2016      0.025     -8.073      0.000      -0.251      -0.152
m1             0.0067      0.003      2.276      0.024       0.001       0.012
tbilrate       1.0191      0.080     12.715      0.000       0.861       1.177
pop           -0.2392      0.067     -3.588      0.000      -0.371      -0.108
==============================================================================
Omnibus:                        9.658   Durbin-Watson:                   1.746
Prob(Omnibus):                  0.008   Jarque-Bera (JB):               20.528
Skew:                          -0.046   Prob(JB):                     3.49e-05
Kurtosis:                       4.616   Cond. No.                     5.67e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 5.67e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

The R-square=0.672 suggests the model fits the data fairly well. I choose these independent variables since their p-val of t < 0.05 in the previous model that contains all independent variables.

The model has adjusted R-square=0.659, which means not all variables are contributing.

The F statistic test the null hypothesis that all of the regression coefficients are equal to zero.  The p-val of F statistic is 2.79e-40 which is < 0.05. This means we can reject the null hypothesis, and the model has some validity in fitting the data.

We set a significance level alpha = 0.05. Here, the contribution of these variables are significiant since their p-val of t < 0.05. Each of these regression coefficients are not equal to zero. In other words, these variables have significant impacts on inflation.

In [30]:
# 2006 first quarter
X1 = pd.DataFrame({'realgdp': [12915.938], 'realinv':[2264.721], 'realgovt':[900.511],'cpi': [200.7],'m1': [1380.5], 'tbilrate':[4.51], 'pop':[298.061]})
lm1.predict(X1)

0    3.603067
dtype: float64

In [31]:
# 2006 second quarter
X2 = pd.DataFrame({'realgdp': [12962.462], 'realinv':[2261.247], 'realgovt':[892.839],'cpi': [202.7],'m1': [1369.2], 'tbilrate':[4.82], 'pop':[298.766]})
lm1.predict(X2)

0    3.867434
dtype: float64

In [32]:
# 2006 third quarter
X3 = pd.DataFrame({'realgdp': [12965.916], 'realinv':[2229.636], 'realgovt':[892.002],'cpi': [201.9],'m1': [1369.4], 'tbilrate':[4.9], 'pop':[299.593]})
lm1.predict(X3)

0    4.498397
dtype: float64

In [33]:
# 2006 fourth quarter
X4 = pd.DataFrame({'realgdp': [13060.679], 'realinv':[2165.966], 'realgovt':[894.404],'cpi': [203.574],'m1': [1373.6], 'tbilrate':[4.92], 'pop':[300.32]})
lm1.predict(X4)

0    5.980303
dtype: float64

In [34]:
# 2007 first quarter
X5 = pd.DataFrame({'realgdp': [13099.901], 'realinv':[2132.609], 'realgovt':[882.766],'cpi': [205.92],'m1': [1379.7], 'tbilrate':[4.95], 'pop':[300.977]})
lm1.predict(X5)

0    6.512911
dtype: float64

In [35]:
# 2007 second quarter
X6 = pd.DataFrame({'realgdp': [13203.977], 'realinv':[2162.214], 'realgovt':[898.713],'cpi': [207.338],'m1': [1370], 'tbilrate':[4.72], 'pop':[301.714]})
lm1.predict(X6)

0    6.007894
dtype: float64

In [36]:
# 2007 third quarter
X7 = pd.DataFrame({'realgdp': [13321.109], 'realinv':[2166.491], 'realgovt':[918.983],'cpi': [209.133],'m1': [1379.2], 'tbilrate':[4], 'pop':[302.509]})
lm1.predict(X7)

0    5.538669
dtype: float64

In [37]:
# 2007 fourth quarter
X8 = pd.DataFrame({'realgdp': [13391.249], 'realinv':[2123.426], 'realgovt':[925.11],'cpi': [212.495],'m1': [1377.4], 'tbilrate':[3.01], 'pop':[303.204]})
lm1.predict(X8)

0    4.986202
dtype: float64

In [38]:
# 2008 first quarter
X9 = pd.DataFrame({'realgdp': [13366.865], 'realinv':[2082.886], 'realgovt':[943.372],'cpi': [213.997],'m1': [1384], 'tbilrate':[1.56], 'pop':[303.803]})
lm1.predict(X9)

0    3.331646
dtype: float64

In [39]:
# 2008 second quarter
X10 = pd.DataFrame({'realgdp': [13415.266], 'realinv':[2026.518], 'realgovt':[961.28],'cpi': [218.61],'m1': [1409.3], 'tbilrate':[1.74], 'pop':[304.483]})
lm1.predict(X10)

0    3.773078
dtype: float64

In [40]:
# 2008 third quarter
X11 = pd.DataFrame({'realgdp': [13324.6], 'realinv':[1990.693], 'realgovt':[991.551],'cpi': [216.889],'m1': [1474.7], 'tbilrate':[1.17],  'pop':[305.27]})
lm1.predict(X11)

0    3.154093
dtype: float64

In [41]:
# 2008 fourth quarter
X12 = pd.DataFrame({'realgdp': [13141.92], 'realinv':[1857.661], 'realgovt':[1007.273],'cpi': [212.174],'m1': [1576.5], 'tbilrate':[0.12], 'pop':[305.952]})
lm1.predict(X12)

0    3.913405
dtype: float64

In [42]:
# 2009 first quarter
X13 = pd.DataFrame({'realgdp': [12925.41], 'realinv':[1558.494], 'realgovt':[996.287],'cpi': [212.671],'m1': [1592.8], 'tbilrate':[0.22], 'pop':[306.547]})
lm1.predict(X13)

0    7.123521
dtype: float64

In [43]:
# 2009 second quarter
X14 = pd.DataFrame({'realgdp': [12901.504], 'realinv':[1456.678], 'realgovt':[1023.528],'cpi': [214.469],'m1': [1653.6], 'tbilrate':[0.18], 'pop':[307.226]})
lm1.predict(X14)

0    8.12451
dtype: float64

In [44]:
# 2009 third quarter
X15 = pd.DataFrame({'realgdp': [12990.341], 'realinv':[1486.398], 'realgovt':[1044.088],'cpi': [216.385],'m1': [1673.9], 'tbilrate':[0.12], 'pop':[308.013]})
lm1.predict(X15)

0    7.675481
dtype: float64

In [45]:
# define true and predicted response values
y_true = [2.6, 3.97, -1.58, 3.3, 4.58, 2.75, 3.45, 6.38, 2.82, 8.53, -3.16, -8.79, 0.94, 3.37, 3.56]
y_pred = [3.603067, 3.867434, 4.498397,  5.980303,  6.512911, 6.007894, 5.538669, 4.986202, 3.331646, 3.773078, 3.154093, 3.913405, 7.123521,  8.12451, 7.675481]

# calculate MAE, MSE, RMSE
print (metrics.mean_absolute_error(y_true, y_pred))
print (metrics.mean_squared_error(y_true, y_pred))
print (np.sqrt(metrics.mean_squared_error(y_true, y_pred)))

3.858478866666667
24.514377697784067
4.951199622089991


MAE measures the average magnitude of the errors in a set of predictions. Here MAE is equal to 3.858 which is not too large. RMSE is the square root of the mean of the squared errors and is interpretable in the "y" units. Lower values of RMSE indicate better fit. Here, the RMSE = 4.951, which means the model can not predict inflation very accurately but is close to the actual data. 